orca/server_metrics.go

```diff
@@ -0,0 +1,271 @@
+	}
+	if sm.CPUUtilization != -1 {
+		ret.CpuUtilization = sm.CPUUtilization
+	}
+	if sm.MemUtilization != -1 {
+		ret.MemUtilization = sm.MemUtilization
+	}
+	if sm.QPS != -1 {
+		ret.RpsFractional = sm.QPS
+	}
+	if sm.EPS != -1 {
+		ret.Eps = sm.EPS
+	}
+	return ret
+}
+
+// merge merges o into sm, overwriting any values present in both.
+func (sm *ServerMetrics) merge(o *ServerMetrics) {
+	if o.CPUUtilization != -1 {
+		sm.CPUUtilization = o.CPUUtilization
+	}
+	if o.MemUtilization != -1 {
+		sm.MemUtilization = o.MemUtilization
+	}
+	if o.QPS != -1 {
+		sm.QPS = o.QPS
+	}
+	if o.EPS != -1 {
+		sm.EPS = o.EPS
+	}
+	mergeMap(sm.Utilization, o.Utilization)
+	mergeMap(sm.RequestCost, o.RequestCost)
+	mergeMap(sm.NamedMetrics, o.NamedMetrics)
+}
+
+func mergeMap(a, b map[string]float64) {
+	for k, v := range b {
+		a[k] = v
+	}
+}
+
+// ServerMetricsRecorder allows for recording and providing out of band server
+// metrics.
+type ServerMetricsRecorder interface {
+	ServerMetricsProvider
+
+	// SetCPUUtilization sets the relevant server metric.
+	SetCPUUtilization(float64)
+	// DeleteCPUUtilization deletes the relevant server metric to prevent it
+	// from being sent.
+	DeleteCPUUtilization()
+
+	// SetMemoryUtilization sets the relevant server metric.
+	SetMemoryUtilization(float64)
+	// DeleteMemoryUtilization deletes the relevant server metric to prevent it
+	// from being sent.
+	DeleteMemoryUtilization()
+
+	// SetQPS sets the relevant server metric.
+	SetQPS(float64)
+	// DeleteQPS deletes the relevant server metric to prevent it from being
+	// sent.
+	DeleteQPS()
+
+	// SetEPS sets the relevant server metric.
+	SetEPS(float64)
+	// DeleteEPS deletes the relevant server metric to prevent it from being
+	// sent.
+	DeleteEPS()
+
+	// SetNamedUtilization sets the relevant server metric.
+	SetNamedUtilization(name string, val float64)
+	// DeleteNamedUtilization deletes the relevant server metric to prevent it
+	// from being sent.
+	DeleteNamedUtilization(name string)
+}
+
+type serverMetricsRecorder struct {
+	mu    sync.Mutex     // protects state
+	state *ServerMetrics // the current metrics
+}
+
+// NewServerMetricsRecorder returns an in-memory store for ServerMetrics and
+// allows for safe setting and retrieving of ServerMetrics.  Also implements
+// ServerMetricsProvider for use with NewService.
+func NewServerMetricsRecorder() ServerMetricsRecorder {
+	return newServerMetricsRecorder()
+}
+
+func newServerMetricsRecorder() *serverMetricsRecorder {
+	return &serverMetricsRecorder{
+		state: &ServerMetrics{
+			CPUUtilization: -1,
+			MemUtilization: -1,
+			QPS:            -1,
+			EPS:            -1,
+			Utilization:    make(map[string]float64),
+			RequestCost:    make(map[string]float64),
+			NamedMetrics:   make(map[string]float64),
+		},
+	}
+}
+
+// ServerMetrics returns a copy of the current ServerMetrics.
+func (s *serverMetricsRecorder) ServerMetrics() *ServerMetrics {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	ret := &ServerMetrics{
```

Comment:
Nit: return the value directly from this line.

Reply:
Done.